### PR TITLE
In 146947449 validate decision allegation interactions

### PR DESCRIPTION
**Pull request by saraOSI**

### CODE REVIEW

File: app/assets/javascripts/components/screenings/AllegationsEditView.jsx [hunk 1]
```diff
@@ -69,6 +76,7 @@ const AllegationsEditView = ({allegations, onSave, onCancel, onChange, alertErro
 AllegationsEditView.propTypes = {
   alertErrorMessage: PropTypes.string,
   allegations: PropTypes.object.isRequired,
+  areAllegationsRequired: PropTypes.bool.isRequired,
```
Comment (NateOSI) [inline app/assets/javascripts/components/screenings/AllegationsEditView.jsx]: should this just be allegationsRequired? or required? given its on the allegations card?

File: spec/features/screening/validations/screening_decision_validations_spec.rb [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  let(:screening) do
+    FactoryGirl.create(
+      :screening,
+      participants: [perpetrator, victim],
+      screening_decision: screening_decision
+    )
+  end
+
+  before do
+    allegation = FactoryGirl.create(
```
Comment (NateOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: should this just be nested under the screening factory above?
```
FactoryGirl.create(
       :screening,
       participants: [perpetrator, victim],
       screening_decision: screening_decision
      allegations: [...]
     )
```

File: app/assets/javascripts/ValidationRules.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import Immutable from 'immutable'
+
+const VALIDATIONS = Object.freeze({
+  decision_card: {
+    screening_decision: [{
+      rule: 'isInvalidIf',
+      condition: (value) => (
+        value === 'promote_to_referral' //&& !this.props.areValidAllegationsPresent
```
Comment (NateOSI) [inline app/assets/javascripts/ValidationRules.js]: what the?????? //&& !this.props.areValidAllegationsPresent

File: app/assets/javascripts/components/screenings/AllegationsEditView.jsx (hunk 1 above)
Comment (saraOSI) [inline app/assets/javascripts/components/screenings/AllegationsEditView.jsx]: `required` is probably a good change here.

File: spec/features/screening/validations/screening_decision_validations_spec.rb (hunk 2 above)
Comment (saraOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: I'm not sure if we can do that in this codebase given there's no real relationship between allegations and screenings.

File: app/assets/javascripts/components/screenings/AllegationsEditView.jsx (hunk 1 above)
Comment (aaronOSI) [inline app/assets/javascripts/components/screenings/AllegationsEditView.jsx]: I vote for `areAllegationsRequired`

File: app/assets/javascripts/components/screenings/DecisionCardView.jsx [hunk 4]
```diff
@@ -33,16 +44,31 @@ export default class DecisionCardView extends React.Component {
     })
   }
 
+  onBlur(fieldName) {
+    const displayErrorsFor = this.state.displayErrorsFor.push(fieldName)
```
Comment (aaronOSI) [inline app/assets/javascripts/components/screenings/DecisionCardView.jsx]: if we track these under something like `this.state.blurredFields` instead of `this.state.displayErrorsFor` we'll be able to leverage it semantically for other things

File: spec/javascripts/components/screenings/decision/DecisionEditViewSpec.jsx [hunk 5]
```diff
@@ -128,10 +141,12 @@ describe('DecisionEditView', () => {
       onChange: jasmine.createSpy(),
       onCancel: jasmine.createSpy(),
       onSave: jasmine.createSpy(),
+      onBlur: jasmine.createSpy(),
```
Comment (aaronOSI) [inline spec/javascripts/components/screenings/decision/DecisionEditViewSpec.jsx]: Is this spy being checked for # of calls / their arguments?

File: spec/support/helpers/screening_helpers.rb [hunk 6]
```diff
@@ -45,15 +45,43 @@ def validate_message_as_user_interacts_with_card(
...
+    stub_request(
+      :get,
+      host_url(ExternalRoutes.intake_api_history_of_involvements_path(screening.id))
+    ).and_return(json_body([].to_json, status: 200))
+  end
+
   def stub_and_visit_edit_screening(screening)
     stub_request(:get, host_url(ExternalRoutes.intake_api_screening_path(screening.id)))
       .and_return(json_body(screening.to_json, status: 200))
+    stub_empty_relationships_for_screening(screening)
```
Comment (aaronOSI) [inline spec/support/helpers/screening_helpers.rb]: Love this. And eventually we can use spec meta tags, like `history: false` to elegantly stub these across all relevant specs. 👍

File: app/assets/javascripts/components/screenings/ScreeningPage.jsx [hunk 7]
```diff
@@ -211,6 +212,7 @@ export class ScreeningPage extends React.Component {
     }
     const {screening, loaded} = this.state
     const mergedScreening = this.mergeScreeningWithEdits(this.state.screeningEdits)
+    const cardErrors = Validator.validateScreening({screening: mergedScreening})
```
Comment (AmanShurafa) [inline app/assets/javascripts/components/screenings/ScreeningPage.jsx]: This introduces dramatic inconsistency in the way we handle errors. This is increases the need to refactor validation. I LIKE IT!!

File: app/assets/javascripts/utils/validator.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 import Immutable from 'immutable'
 import _ from 'lodash'
 import moment from 'moment'
+import VALIDATIONS from 'ValidationRules'
```
Comment (AmanShurafa) [inline app/assets/javascripts/utils/validator.js]: Don't like tying validator for 1 specific set of rules. I would rather pass it. This is debatable. I guess we can keep it this way for now and change it if there is a need for that later.

File: spec/javascripts/components/screenings/decision/DecisionEditViewSpec.jsx (hunk 5 above)
Comment (saraOSI) [inline spec/javascripts/components/screenings/decision/DecisionEditViewSpec.jsx]: Nope. I can add a test for that.

File: app/assets/javascripts/ValidationRules.js (hunk 3 above)
Comment (saraOSI) [inline app/assets/javascripts/ValidationRules.js]: Artifact of a refactor. It's cleaned up in the next commit. I wanted to make sure that I didn't lose track of the requirements as I was moving code around.

File: spec/javascripts/components/screenings/decision/DecisionEditViewSpec.jsx (hunk 5 above)
Comment (saraOSI) [inline spec/javascripts/components/screenings/decision/DecisionEditViewSpec.jsx]: Added.

File: spec/support/helpers/screening_helpers.rb [hunk 9]
```diff
@@ -45,15 +45,43 @@ def validate_message_as_user_interacts_with_card(
...
+  def stub_and_visit_show_screening(screening)
+    stub_request(:get, host_url(ExternalRoutes.intake_api_screening_path(screening.id)))
+      .and_return(json_body(screening.to_json, status: 200))
+    stub_empty_relationships_for_screening(screening)
+    stub_empty_history_for_screening(screening)
+
+    visit screening_path(id: screening.id)
+
+    # TODO: remove this once we can consistently have a fresh page for these specs
+    page.evaluate_script('window.location.reload()')
```
Comment (saraOSI) [inline spec/support/helpers/screening_helpers.rb]: Apparently we do. :-(

File: spec/features/screening/validations/screening_decision_validations_spec.rb [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+          expect(page).not_to have_content(error_message)
+          click_button 'Cancel'
+        end
+
+        within '#decision-card.show' do
+          expect(page).not_to have_content(error_message)
+          click_link 'Edit'
+        end
+
+        stub_screening_put_request_with_anything_and_return(
```
Comment (saraOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: 👍

File: spec/features/screening/validations/screening_decision_validations_spec.rb [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+          select 'Promote to referral', from: 'Screening Decision'
+          blur_field
+          expect(page).to have_content(error_message)
+          select 'Screen out', from: 'Screening Decision'
+          blur_field
+          expect(page).not_to have_content(error_message)
+        end
+      end
+
+      scenario 'Adding and removing allegations shows or hides error message' do
```
Comment (saraOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: The previous two scenarios test that the errors display/don't display depending on what you select within the decision card. This scenario specifically addresses that the errors shows/hides when the user interacts with the allegations, which are in a separate card. This test was failing until I pulled the logic to the screening level, so I think it's a necessary test.

File: app/assets/javascripts/components/screenings/DecisionCardView.jsx (hunk 4 above)
Comment (saraOSI) [inline app/assets/javascripts/components/screenings/DecisionCardView.jsx]: We're pushing items into this list on card save, as well, since we always display errors in save mode. Per our discussion, leaving this as-is until we have a need to track which fields have been blurred vs saved, etc.

File: spec/features/screening/validations/screening_decision_validations_spec.rb [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        within '#decision-card.show' do
+          expect(page).not_to have_content(error_message)
+        end
+      end
+    end
+
+    context 'Screening decision is already set to promote to referral' do
+      let(:screening_decision) { 'promote_to_referral' }
+
+      scenario 'User does not see error messages on page load' do
```
Comment (saraOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: Good catch.

**Comment by saraOSI** [comment]
Addressed feedback and got verbal approval from Aaron and Nate before merging.

File: spec/javascripts/ScreeningValidatorSpec.js [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        screening = Immutable.fromJS({some_field: ''})
+        fieldValidations = Immutable.fromJS({some_field: [
+          {rule: 'isRequired', message: 'Please enter a field.'},
+        ]})
+        expect(new ScreeningValidator({}).validateAllFields({screening: screening, fieldValidations: fieldValidations}).get('some_field').toJS()).toContain('Please enter a field.')
+      })
+    })
+  })
+
+  describe('areValidAllegationsPresent', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/ScreeningValidatorSpec.js]: Can we add a test for allegations being an empty list?
